rust/src/storage/s3/store.rs

```diff
@@ -0,0 +1,658 @@
+                .map_err(|_| StorageError::S3Error)?;
+        }
+        Ok(())
+    }
+
+    /// Create all [`Buckets`]
+    pub async fn create_buckets(&self) -> Result<(), StorageError> {
+        for bucket in self.buckets.names() {
+            let resp = self.create_bucket(bucket).await;
+
+            if let Err(RusotoError::Service(CreateBucketError::BucketAlreadyExists(_)))
+            | Err(RusotoError::Service(CreateBucketError::BucketAlreadyOwnedByYou(_))) = resp
+            {
+                continue;
+            } else {
+                return Err(StorageError::S3Error);
+            }
+        }
+        Ok(())
+    }
+}
+
+// private methods
+impl S3Store {
+    /// Upload an object to the given bucket.
+    async fn upload(
+        &self,
+        bucket: &str,
+        key: &str,
+        data: Vec<u8>,
+    ) -> Result<PutObjectOutput, RusotoError<PutObjectError>> {
+        let req = PutObjectRequest {
+            bucket: bucket.to_string(),
+            key: key.to_string(),
+            body: Some(StreamingBody::from(data)),
+            ..Default::default()
+        };
+
+        self.s3_client.put_object(req).await
+    }
+
+    /// Download an object from the given bucket.
+    async fn download_object(
+        &self,
+        bucket: &str,
+        key: &str,
+    ) -> Result<GetObjectOutput, RusotoError<GetObjectError>> {
+        let req = GetObjectRequest {
+            bucket: bucket.to_string(),
+            key: key.to_string(),
+            ..Default::default()
+        };
+
+        self.s3_client.get_object(req).await
+    }
+
+    // Get the content of the given object
+    async fn unpack_object(&self, object: GetObjectOutput) -> Option<Vec<u8>> {
+        if let Some(body) = object.body {
+            let mut content = Vec::new();
+            // TODO handle error
+            body.into_async_read().read_to_end(&mut content).await;
+            Some(content)
+        } else {
+            None
+        }
+    }
+
+    // Create a new bucket with the given bucket name
+    async fn create_bucket(
+        &self,
+        bucket: &str,
+    ) -> Result<CreateBucketOutput, RusotoError<CreateBucketError>> {
+        let req = CreateBucketRequest {
+            bucket: bucket.to_string(),
+            ..Default::default()
+        };
+        self.s3_client.create_bucket(req).await
+    }
+
+    async fn clear_bucket(&self, bucket: &str) -> Result<(), StorageError> {
+        let mut continuation_token: Option<String> = None;
+
+        loop {
+            let list_obj_resp = self
+                .list_objects(bucket, continuation_token)
+                .await
+                .map_err(|_| StorageError::S3Error)?;
+
+            if let Some(identifiers) = self.unpack_object_identifier(&list_obj_resp) {
+                self.delete_objects(bucket, identifiers)
+                    .await
+                    .map_err(|_| StorageError::S3Error)?;
+            } else {
+                break Ok(());
```

Comment:
nit: I know this is a matter of taste, I suggest we just `break;` and return `Ok(())` at the end of the function. This confused me for a second.